pkg/planner/optimize.go

```
@@ -66,28 +67,38 @@ func containUsePlanCacheHintInSQLOrBinding(sctx sessionctx.Context, stmt ast.Stm
-		!sctx.GetSessionVars().DisableTxnAutoRetry || // txn-auto-retry
-		sctx.GetSessionVars().InMultiStmts { // in multi-stmt
+		!vars.DisableTxnAutoRetry || // txn-auto-retry
+		vars.InMultiStmts || // in multi-stmt
+		(stmtCtx.InExplainStmt && stmtCtx.ExplainFormat != types.ExplainFormatPlanCache) { // in explain internal
 		return nil, nil, false, nil
 	}
-	if sctx.GetSessionVars().PlanCacheStrategy == vardef.TiDBPlanCacheStrategyHintOnly &&
+	if vars.PlanCacheStrategy == vardef.TiDBPlanCacheStrategyHintOnly &&
 		!containUsePlanCacheHintInSQLOrBinding(sctx, stmt) {
 		if !isExplain && stmtCtx.InExplainStmt && stmtCtx.ExplainFormat == types.ExplainFormatPlanCache {
 			stmtCtx.AppendWarning(errors.NewNoStackErrorf("skip non-prepared plan-cache: %s", nonPreparedPlanCacheHintOnlyNoHintReason))
```

golangci-lint (2.12.2)

[error] The line is 127 characters long, which exceeds the maximum of 120 characters.

(lll)